packages/pkl.lua/tests/lua.pkl

```diff
@@ -0,0 +1,367 @@
+    new lua.Renderer {
+      converters {
+        ["*"] = (it) -> if (it is Int) it + 1 else it
+      }
+    }.renderDocument(new Dynamic { 1; a = 1; b { 1; c = 1; d { e = 1 } } }) == new lua.Renderer {}.renderDocument(new Dynamic { 1; a = 2; b { 1; c = 2; d { e = 2 } } })
+    new lua.Renderer {
+      converters {
+        ["a.b"] = (it) -> it + 1
+      }
+    }.renderDocument(new Dynamic { a { b = 1; c { b = 1 } }; b = 1; c { a { b = 1 } } }) == new lua.Renderer {}.renderDocument(new Dynamic { a { b = 2; c { b = 1 } }; b = 1; c { a { b = 2 } } })
+    new lua.Renderer {
+      converters {
+        ["^a.b"] = (it) -> it + 1
+      }
+    }.renderDocument(new Dynamic { a { b = 1; c { b = 1 } }; b = 1; c { a { b = 1 } } }) == new lua.Renderer {}.renderDocument(new Dynamic { a { b = 2; c { b = 1 } }; b = 1; c { a { b = 1 } } })
+    new lua.Renderer {
+      converters {
+        ["a"] = (it) -> new Dynamic { b = it }
+        ["b"] = (it) -> new Dynamic { c = it }
+      }
+    }.renderDocument(new Dynamic { a = 1 }) == new lua.Renderer {}.renderDocument(new Dynamic { a { b { c = 1 } } })
+    new lua.Renderer {
+      converters {
+        ["[*]"] = (it) -> if (it is Int) it + 1 else it
+      }
+    }.renderDocument(new Mapping { ["a"] = 1; ["b"] = new Mapping { ["c"] = 1 } }) == new lua.Renderer {}.renderDocument(new Dynamic { a = 2; b { c = 2 } })
+    new lua.Renderer {
+      converters {
+        ["[*]"] = (it) -> if (it is Int) it + 1 else it
+      }
+    }.renderDocument(new Dynamic { a = 1; 1; new Listing { 1; 2 }; 2 }) == new lua.Renderer {}.renderDocument(new Dynamic { a = 1; 2; new Listing { 2; 3 }; 3 })
+    new lua.Renderer {
+      converters {
+        ["[a]"] = (it) -> it + 1
+      }
+    }.renderDocument(new Dynamic { a = 1; map = new Mapping { ["a"] = 1 } }) == new lua.Renderer {}.renderDocument(new Dynamic { a = 1; map { ["a"] = 2 } })
+    // check that path like "[1]" matches string key "1" and not list idx or int key or funky property
+    let (source = new Dynamic { 1; 1; new Dynamic { 1; 1 }; new Listing { 1; 1 }; new Mapping { [1] = 1; ["1"] = 1 }; new Dynamic { `[1]` = 1; `["1"]` = 1 } })
+    let (expected = new Dynamic { 1; 1; new Dynamic { 1; 1 }; new Listing { 1; 1 }; new Mapping { [1] = 1; ["1"] = 2 }; new Dynamic { `[1]` = 1; `["1"]` = 1 } })
+    new lua.Renderer {
+      converters {
+        ["[1]"] = (it) -> it + 1
+      }
+    }.renderDocument(source) == new lua.Renderer {}.renderDocument(expected)
+    // check for string vs int key on Dynamic too; note that Dynamic treats all string keys as properties, so the converter shouldn't run
+    let (source = new Dynamic { [1] = 1; ["1"] = 1; nest { [1] = 1; ["1"] = 1 } })
+    let (expected = new Dynamic { [1] = 1; ["1"] = 1; nest { [1] = 1; ["1"] = 1 } })
+    new lua.Renderer {
+      converters {
+        ["[1]"] = (it) -> it + 1
+      }
+    }.renderDocument(source) == new lua.Renderer {}.renderDocument(expected)
```

Comment:
This case isn't working right now:

```groovy
import "@lua/lua"

foo {
  ["bar"] = 5
}

output {
  renderer = new lua.Renderer {
    converters {
      ["foo[bar]"] = (it) -> it + 1
    }
  }
}
```

I would expect this to produce:

```
foo = { bar = 6; }
```

Reply:
Also, you might want to consider writing these as examples, rather than facts, because they help you diff output when things fail.

Reply:
This is a known failure case, it's due to a language limitation where Pkl code cannot distinguish between properties and entries when iterating over a `Dynamic`. This limitation is mentioned in the doc comment on the `converters` property (and is the reason why that property is redeclared).

Reply:
Gotcha, sounds good!